src/analysis/analysis-service.js

```diff
@@ -0,0 +1,151 @@
+var _ = require('underscore');
+var Analysis = require('./analysis-model');
+var camshaftReference = require('./camshaft-reference');
+var LayerTypes = require('../geo/map/layer-types.js');
+
+var AnalysisService = function (opts) {
+  opts = opts || {};
+  if (!opts.vis) {
+    throw new Error('vis option is required');
+  }
+
+  this._vis = opts.vis;
+  this._apiKey = opts.apiKey;
+  this._authToken = opts.authToken;
+  this._camshaftReference = opts.camshaftReference || camshaftReference; // For testing purposes
+
+  this._analysisNodes = {};
+};
+
+/**
+  * Recursively generates a graph of analyses and returns the "root" node. Each node
+  * may have one or more "source" params pointing to another node. If a node had been created
+  * already, this method updates the attributes of the existing node. New nodes are added to
+  * the collection of analyses that has been injected.
+  */
+AnalysisService.prototype.analyse = function (analysisDefinition) {
+  analysisDefinition = _.clone(analysisDefinition);
+  var analysis = this.findNodeById(analysisDefinition.id);
+  var analysisAttrs = this._getAnalysisAttributesFromAnalysisDefinition(analysisDefinition);
+
+  if (analysis) {
+    analysis.set(analysisAttrs);
+  } else {
+    if (this._apiKey) {
+      analysisAttrs.apiKey = this._apiKey;
+    }
+    if (this._authToken) {
+      analysisAttrs.authToken = this._authToken;
+    }
+    analysis = new Analysis(analysisAttrs, {
+      camshaftReference: this._camshaftReference,
+      vis: this._vis
+    });
+
+    this._analysisNodes[analysisDefinition.id] = analysis;
+    analysis.bind('destroy', this._onAnalysisRemoved, this);
+  }
+
+  return analysis;
+};
+
+/**
+ * This function applies recursively the function analyse to the sourceNodes in the analysisDefinition
```

Comment:
The description is wrong.

> This function is used to iterate over the analysis graph. It uses the camshaft reference to extract those parameters which are analysis nodes. And call `analyse` on them.

Maybe we could add a comment remaining this wont be needed if we split `parameters` and `inputs` in the analysis model